helix-core/src/main/java/org/apache/helix/spectator/RoutingTableProvider.java

```diff
@@ -376,50 +575,104 @@ public Object removeRoutingTableChangeListener(
    * @return
    */
   public Collection<LiveInstance> getLiveInstances() {
-    return _routingTableRef.get().getLiveInstances();
+    // Since line instances will be the same across all _routingTableRefMap, here one of the keys
+    // will be used without considering PropertyType
+    String key = null;
+    Iterator<String> iter = _routingTableRefMap.keySet().iterator();
+    if (iter.hasNext()) {
+      key = iter.next();
+    }
+    if (key == null) {
+      throw new HelixException("There is no key available in this RoutingTableProvider.");
+    }
+    return _routingTableRefMap.get(key).get().getLiveInstances();
   }
 
   /**
    * Return all instance's config in this cluster.
    * @return
    */
   public Collection<InstanceConfig> getInstanceConfigs() {
-    return _routingTableRef.get().getInstanceConfigs();
+    // Since line instances will be the same across all _routingTableRefMap, here one of the keys
+    // will be used without considering PropertyType
+    String key = null;
+    Iterator<String> iter = _routingTableRefMap.keySet().iterator();
+    if (iter.hasNext()) {
+      key = iter.next();
+    }
+    if (key == null) {
+      throw new HelixException("There is no key available in this RoutingTableProvider.");
+    }
+    return _routingTableRefMap.get(key).get().getInstanceConfigs();
   }
 
   /**
-   * Return names of all resources (shown in ExternalView) in this cluster.
+   * Return names of all resources (shown in ExternalView or CustomizedView) in this cluster.
    */
   public Collection<String> getResources() {
-    return _routingTableRef.get().getResources();
+    String key = getRoutingTableKey(DEFAULT_PROPERTY, DEFAULT_TYPE);
+    return _routingTableRefMap.get(key).get().getResources();
+  }
+
+  public Collection<String> getResources(PropertyType propertyType, String type) {
+    String key = getRoutingTableKey(propertyType.name(), type);
+    return _routingTableRefMap.get(key).get().getResources();
+  }
+
+
+  private String getRoutingTableKey(String propertyTypeName, String customizedStateType) {
+    if (propertyTypeName.equals(DEFAULT_PROPERTY) && customizedStateType.equals(DEFAULT_TYPE)) {
```

Comment:
Do we have DEFAULT_PROPERTY but none DEFAULT_TYPE case? If not, shall we just ignore the state type if DEFAULT_PROPERTY is used?

Reply:
Makes sense. Done.